caikit/runtime/__main__.py

```diff
@@ -0,0 +1,58 @@
+# Standard
+import signal
+
+# First Party
+import alog
+
+# Local
+from caikit.config.config import get_config
+from caikit.runtime.grpc_server import RuntimeGRPCServer
+from caikit.runtime.http_server import RuntimeHTTPServer
+
+log = alog.use_channel("RUNTIME-MAIN")
+
+
+def main():
+    _grpc_server = None
+    _http_server = None
+
+    def interrupt(signal_, _stack_frame):
+        log.info(
+            "<RUN87630120I>",
+            "Caikit Runtime received interrupt signal %s, shutting down",
+            signal_,
+        )
+        if _grpc_server:
+            _grpc_server.stop()
+        if _http_server:
+            _http_server.stop()
+
+    # NOTE: signal function can only be called from main thread of the main
+    # interpreter. If this function is called from a thread (like in tests)
+    # then signal handler cannot be used. Thus, we will only have real
+    # termination_handler when this is called from the __main__.
+
+    signal.signal(signal.SIGINT, interrupt)
```

Comment:
This is interesting. I was surprised to learn that signals can only be bound once. As written this will work because the individual servers call `signal.signal` in their inits and these calls come afterwards to [override the previously registered handlers](https://docs.python.org/3/library/signal.html#signal.signal). This seems a bit confusing and relies on knowledge that `signal.signal` has overwrite behavior. I think it might be cleaner to NOT call `signal.signal` in the individual server inits and instead do those individual-server registrations in the `main` functions of those modules so that the pattern is "the code defines the `interrupt` method, the `main` registers the signals.

Reply:
Hmm. That's how I have it right now? The individual servers call `server._intercept_interrupt_signal()` within their `main` functions, and I made it like that to avoid running it when we run the `__main__` script (which just calls their init functions).